integration/tests/handshake.test.ts

```diff
@@ -885,3 +886,524 @@ test.describe('Client handshake @generic', () => {
+    res.end();
+  });
+
+  test.beforeAll('setup local jwks server', async () => {
+    // Start the jwks server
+    await new Promise<void>(resolve => jwksServer.listen(0, resolve));
+  });
+
+  test.afterAll('setup local Clerk API mock', async () => {
+    return new Promise<void>(resolve => jwksServer.close(() => resolve()));
+  });
+
+  const start = async (orgSyncOptions: OrganizationSyncOptions): Promise<Application> => {
+    const env = appConfigs.envs.withEmailCodes
+      .clone()
+      .setEnvVariable('private', 'CLERK_API_URL', `http://localhost:${jwksServer.address().port}`);
+
+    const middlewareFile = `import { authMiddleware } from '@clerk/nextjs/server';
+    // Set the paths that don't require the user to be signed in
+    const publicPaths = ['/', /^(\\/(sign-in|sign-up|app-dir|custom)\\/*).*$/];
+    export const middleware = (req, evt) => {
+      return authMiddleware({
+        publicRoutes: publicPaths,
+        publishableKey: req.headers.get("x-publishable-key"),
+        secretKey: req.headers.get("x-secret-key"),
+        proxyUrl: req.headers.get("x-proxy-url"),
+        domain: req.headers.get("x-domain"),
+        isSatellite: req.headers.get('x-satellite') === 'true',
+        signInUrl: req.headers.get("x-sign-in-url"),
+
+        // Critical
+        organizationSyncOptions: ${JSON.stringify(orgSyncOptions)}
+
+      })(req, evt)
+    };
+    export const config = {
+      matcher: ['/((?!.+\\.[\\w]+$|_next).*)', '/', '/(api|trpc)(.*)'],
+    };
+    `;
+
+    const app = await appConfigs.next.appRouter
+      .clone()
+      .addFile('src/middleware.ts', () => middlewareFile)
+      .commit();
+
+    await app.setup();
+    await app.withEnv(env);
+    await app.dev();
+    return app;
+  };
+
+  const end = async (app: Application): Promise<void> => {
+    await app.teardown();
+  };
+
+  type testCase = {
+    name: string;
+    when: when;
+    then: then;
+  };
+  type when = {
+    // With this initial state...
+    initialAuthState: 'active' | 'expired' | 'early';
+    initialSessionClaims: Map<string, string>;
+
+    // When the customer app specifies these orgSyncOptions to middleware...
+    orgSyncOptions: OrganizationSyncOptions;
+
+    // And a request arrives to the app at this path...
+    appRequestPath: string;
+
+    // With a token specified in...
+    tokenAppearsIn: 'header' | 'cookie';
+  };
+
+  type then = {
+    // A handshake should (or should not) occur:
+    expectStatus: number;
+
+    // The middleware should redirect to fapi with this query param value:
+    fapiOrganizationIdParamValue: string | null;
+  };
+
+  const cookieAuthCases: testCase[] = [
+    // ---------------- Session active vs expired tests ----------------
+    // Note: it would be possible to run _every_ test with both active and expired initial states
+    //       and expect the same results, but we're avoiding that to save some test execution time.
+    {
```

Comment:
There are a lot of test cases here, but there could be more! The tests here include every (anticipated) permutation of every input parameter in isolation, but not every combination of permutations (to keep a reasonable execution time)

See this (clerk internal) [spreadsheet](https://docs.google.com/spreadsheets/d/15OVcrHz_tj1K9XSfgPDzEQbjQLqn7XotJaZdZbLrLI4/edit?usp=sharing) that I used to track which permutations are tested.